astropy/stats/funcs.py

```diff
@@ -759,22 +768,43 @@ def median_absolute_deviation(a, axis=None):
     # See https://github.com/numpy/numpy/issues/7330 why using np.ma.median
     # for normal arrays should not be done (summary: np.ma.median always
     # returns an masked array even if the result should be scalar). (#4658)
-    if isinstance(a, np.ma.MaskedArray):
+    if isinstance(data, np.ma.MaskedArray):
         func = np.ma.median
+        if ignore_nan:
+            data = np.ma.masked_where(np.isnan(data), data)
+    elif ignore_nan:
+        if NUMPY_LT_1_10 or axis is None:
+            # older versions don't have nanmedian, but nanmedian
+            # may be faster than np.ma.median *if* axis is None
+            # https://gist.github.com/MSeifert04/1551e80e1aa85b48c9191e727c38f9a9
+            data = np.ma.masked_where(np.isnan(data), data)
+            func = np.ma.median
+        else:
+            func = np.nanmedian
     else:
         func = np.median
 
-    a = np.asanyarray(a)
-    a_median = func(a, axis=axis)
+    if not ignore_nan and np.any(np.isnan(data)) and NUMPY_LT_1_10:
```

Comment:
Yes, it's a lot of overhead.  However, I think it's somewhat acceptable to add this overhead for old numpy versions, because you could get a very wrong result if you don't know this.  I will change the order, though - that's obviously right.